benchmarks/capi/GEOSPreparedContainsPerfTest.cpp

```diff
@@ -0,0 +1,104 @@
+/**********************************************************************
+ *
+ * GEOS - Geometry Engine Open Source
+ * http://geos.osgeo.org
+ *
+ * Copyright (C) 2019 Daniel Baston <dbaston@gmail.com>
+ *
+ * This is free software; you can redistribute and/or modify it under
+ * the terms of the GNU Lesser General Public Licence as published
+ * by the Free Software Foundation.
+ * See the COPYING file for more information.
+ *
+ **********************************************************************/
+
+#include <geos/profiler.h>
+#include <geos_c.h>
+
+#include <geos/geom/Coordinate.h>
+#include <geos/geom/GeometryFactory.h>
+#include <geos/geom/Polygon.h>
+#include <geos/geom/util/SineStarFactory.h>
+
+#include <algorithm>
+#include <random>
+#include <vector>
+#include <memory>
+
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+class GEOSPreparedContainsPerfTest {
+
+public:
+    void test(const GEOSGeometry* g, size_t num_points) {
+        using namespace geos::geom;
+
+        double xmin, xmax, ymin, ymax;
+
+        GEOSGeom_getXMin(g, &xmin);
+        GEOSGeom_getXMax(g, &xmax);
+        GEOSGeom_getYMin(g, &ymin);
+        GEOSGeom_getYMax(g, &ymax);
+
+        std::default_random_engine e(12345);
+        std::uniform_real_distribution<> xdist(xmin, xmax);
+        std::uniform_real_distribution<> ydist(ymin, ymax);
+
+        std::vector<Coordinate> coords(num_points);
+        std::generate(coords.begin(), coords.end(), [&xdist, &ydist, &e]() {
+            return Coordinate(xdist(e), ydist(e));
+        });
+
+        geos::util::Profile sw("GEOSPreparedContains");
+        sw.start();
+
+        size_t hits = 0;
+        auto prep = GEOSPrepare(g);
+        for (const auto& c : coords) {
+            auto seq = GEOSCoordSeq_create(1, 2);
```

Comment:
@dbaston Any thoughts on the perf hit from creating a new CoordSeq and Point each time? Is it possible to reuse the same CS and Point object, just resetting the XY each time?  Seems like a bit of a hack at the moment, but if it provides a decent improvement then it would be worth formalizing this pattern.  (This I think addresses the issue that PDAL had with using GEOS for PIP).

Reply:
@dr-jts I don't have numbers in front of me ATM, but the hit is significant. I don't think we allow changing a CoordinateSequence after it has been given to a Geometry (for one, wouldn't it make a lazy-constructed Envelope incorrect?). There are two ways I'd like to experiment with on this (both could possibly be adopted):

1) Create a new `CoordinateSequence` implementation that is backed by `std::array` for small values of `n`. This should compile to the equivalent of a single `Coordinate` while providing the hooks we need for various parts of our API (`size()`, filter methods, etc.)
2) Expose some new C API functions along the lines of `GEOSPreparedContainsXY(GEOSPreparedGeometry*, double x, double y)` that avoid creating Point geometries altogether.

Reply:
There is the `Geometry.geometryChanged` method that allows invalidating the envelope after a coordinate change.   This pattern is supported in JTS, although it's only used in one place.  

Generally it's encouraged to treat Geometrys as immutable.  But if it's needed for performance this can probably be allowed.  And in this case the mutation can be hidden inside the GEOS API I think, so is less risky.  

Can you hack the performance test on this PR to see if this works, and get some idea if there's a perf improvement?

Reply:
For that to work, the coordinate change would have to be initiated by the Geometry, right? (with a `filter` method, for example). In the C API we can only modify a coordinate by our handle on the `CoordinateSequence`. And the `CoordinateSequence` does not know which geometry owns it, if any.

Reply:
> For that to work, the coordinate change would have to be initiated by the Geometry, right? (with a `filter` method, for example).

Nope, this was designed to support external notification to a Geometry that its coordinates changed.

 In the C API we can only modify a coordinate by our handle on the `CoordinateSequence`. And the `CoordinateSequence` does not know which geometry owns it, if any.

My thinking was that for the prepared PIP use case the client could create a CS and the Point using it, and then reuse both by changing the ordinates of the CS.  Maybe not supported by the C API currently, but would this be possible in C++ now?  

Calling `geometryChanged` after setting the ordinates is the Right Thing To Do.  But really the goal of the changes in this PR are to avoid looking at the Point envelope entirely, right?!  So maybe not strictly necessary?

If this checks out as working, then could expose this in some way via C API.  Perhaps by returning a handle to a context created by C API, which keeps the Point and CS objects around until disposed of.

Reply:
Oh, yes, you could make it work in C++ if you manually called `geometryChanged` after changing the `CoordinateSequence`. To do it in C you'd need to expose `geometryChanged` in the C API. My point was that the `CoordinateSequence` can't automatically notify the geometry that it was changed. It seems like a pattern that's better used within the library than as part of the public API.

Reply:
Fair enough.  So is it possible to work up a C API call(s) which allow reusing Point and CS for say PIP calls?

Reply:
Yes, though I'd tackle that separately from this PR. I'm inclined to start by making point creation more efficient and looking at reuse strategies only if it remains a bottleneck.

Reply:
Makes sense.  Great work on this as it is!

Reply:
Thanks. By the way, this gives a picture of point construction costs relative to the PIP call.

![image](https://user-images.githubusercontent.com/6318931/59074887-c224a280-889b-11e9-9ee2-179fed905d3f.png)

Reply:
So Point construction is 5%, and running PIP is 25%?  Which means not much point in further optimization of Point construction?

Reply:
Well, 5% to create a coordseq, 3% to create a point, 5% to destroy a point....all in all, 13-14% total of total runtime.

The PIP test itself is 25% of total program runtime, so constructing a point is about half as expensive as the PIP test. I'd say there's room for improvement :)